cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -126,6 +131,84 @@ public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws I
   }
 
+  @Override
+  public CloseableIterator<StructuredRow> scan(Range keyRange, int limit, Field<?> orderByField, SortOrder sortOrder)
+    throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Scan range {} with limit {}, sort field {} and sort order {}", schema.getTableId(),
+              keyRange, limit, orderByField, sortOrder);
+    fieldValidator.validateField(orderByField);
+    if (!schema.isIndexColumn(orderByField.getName())) {
+      throw new InvalidFieldException(schema.getTableId(), orderByField.getName(), "is not an indexed column");
+    }
+    ScannerIterator scannerIterator = new ScannerIterator(getScanner(keyRange), schema);
+    LimitIterator limitIterator = new LimitIterator(Collections.singleton(scannerIterator).iterator(),
+                                                    limit);
+    PriorityQueue<StructuredRow> sorted = new PriorityQueue<>(sortOrder.equals(SortOrder.ASC) ?
+                                                                (row1, row2) ->
+                                                                  compareRowFields(row1, row2, orderByField)
+                                                                : (row1, row2) ->
+      compareRowFields(row2, row1, orderByField));
+
+    while (limitIterator.hasNext()) {
+      StructuredRow row = limitIterator.next();
+      sorted.offer(row);
+    }
+    // return an iterator for the sorted elements
+    return new AbstractCloseableIterator<StructuredRow>() {
+      @Override
+      protected StructuredRow computeNext() {
+        while (!sorted.isEmpty()) {
+          return sorted.poll();
+        }
+        return endOfData();
+      }
+
+      @Override
+      public void close() {
+        // no-op
+      }
+    };
+  }
+
+
+  private int compareRowFields(StructuredRow row1, StructuredRow row2, Field<?> orderByField)
+    throws InvalidFieldException {
+    switch (orderByField.getFieldType()) {
+      case INTEGER:
+        return Objects.compare(row1.getInteger(orderByField.getName()), row2.getInteger(orderByField.getName()),
+                               Integer::compare);
+      case LONG:
+        return Objects.compare(row1.getLong(orderByField.getName()), row2.getLong(orderByField.getName()),
+                               Long::compare);
+      case FLOAT:
+        return Objects.compare(row1.getFloat(orderByField.getName()), row2.getFloat(orderByField.getName()),
+                               Float::compare);
+      case DOUBLE:
+        return Objects.compare(row1.getDouble(orderByField.getName()), row2.getDouble(orderByField.getName()),
+                               Double::compare);
+      case STRING:
+        return Objects.compare(row1.getString(orderByField.getName()), row2.getString(orderByField.getName()),
+                               String::compareTo);
+      case BYTES:
+        return Arrays.equals(row1.getBytes(orderByField.getName()), row2.getBytes(orderByField.getName())) ? 0 : -1;
```

Comment:
This will not sort properly. Compare needs to return -1 if lower, 0 if equal, and 1 if greater. The Bytes utility class has a comparator you can use.

Reply:
I think it's this: `return new Bytes.ByteArrayComparator()
          .compare(row1.getBytes(orderByField.getName()), row2.getBytes(orderByField.getName()));`?